llmfoundry/models/llm_embed/modeling_llm_embed.py

```diff
@@ -0,0 +1,439 @@
+
+        # Set the vector representation to either be the average of all the individual token vectors,
+        # or the EOS token at the end of the sequence
+        self.vector_representation = contrastive_config_obj.vector_representation
+        self.normalize_output = contrastive_config_obj.normalize_output
+
+        self.step_size = contrastive_config_obj.pos_step_size
+        self.gather_in_batch_negatives = contrastive_config_obj.gather_in_batch_negatives
+        self.use_legacy_gradient_passthrough = contrastive_config_obj.use_legacy_gradient_passthrough
+        self.n_active_params = sum(p.numel() for p in self.parameters())
+        if loss_fn == 'fused_crossentropy':
+            try:
+                from flash_attn.losses.cross_entropy import \
+                    CrossEntropyLoss as FusedCrossEntropyLoss
+
+                self.loss_fn = FusedCrossEntropyLoss(ignore_index=-100)
+            except:
+                raise ValueError(
+                    'Fused Cross Entropy is not installed. Either (1) have a CUDA-compatible GPU '
+                    +
+                    'and `pip install .[gpu]` if installing from source or `pip install xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v1.0.3#subdirectory=csrc/xentropy` '
+                    +
+                    'if installing from pypi, or (2) set your config model.loss_fn=torch_crossentropy.',
+                )
+        elif loss_fn == 'torch_crossentropy':
+            self.loss_fn = nn.CrossEntropyLoss(ignore_index=-100)
+        else:
+            raise ValueError(
+                f'Specified loss_fn={loss_fn} not recognized. `loss_fn` must be one of [`fused_crossentropy`, `torch_crossentropy`].',
+            )
+
+        self.infonce_process_group = None
+        if contrastive_config_obj.infonce_process_group_size is not None:
+            pg_size = contrastive_config_obj.infonce_process_group_size
+            self.infonce_process_group = create_set_process_group(pg_size)
+
+    def construct_model(self):
+        if self.pretrained_model_name_or_path:
+            model_class = registry.models.get('hf_causal_lm')
+            model_class = cast(type[ComposerHFCausalLM], model_class)
+            model = model_class.build_inner_model(
+                pretrained=True,
+                pretrained_model_name_or_path=self.
+                pretrained_model_name_or_path,
+                pretrained_lora_id_or_path=self.pretrained_lora_id_or_path,
+                trust_remote_code=self.trust_remote_code,
+                init_device=self.init_device,
+                use_flash_attention_2=self.use_flash_attention_2,
+                use_auth_token=self.use_auth_token,
+                config_overrides=self.config_overrides or {},
+                load_in_8bit=self.load_in_8bit,
+                **self.kwargs,
+            )
+        else:
+            model = MPTForCausalLM(MPTConfig(**self.kwargs))
```

Comment:
noting for later, this shouldn't really be an MPTForCausalLM since we don't use the lm head at all. it should just be the backbone model.